### PR TITLE
Do not assume dss never decreases.

### LINKED ISSUE #802: Do not assume dss break never decreases
```{chunk,extent}_dss_max_update()``` assumes that the dss break will never decrease, but the application could conceivably call ```sbrk()``` twice, the second time with a negative value, and this could cause an infinite loop in jemalloc.  This is an improbable (almost implausible) failure mode, but should be fixed nonetheless.

This issue should be fixed on both the stable-4 and dev branches.

### PR DESCRIPTION
**Pull request by xhochy**
An sbrk() caller outside jemalloc can decrease the dss, so add a
separate atomic boolean to explicitly track whether jemalloc is
concurrently calling sbrk(), rather than depending on state outside
jemalloc's full control.

Fixes #802 for stable-4

This is a backport of https://github.com/jemalloc/jemalloc/commit/196a53c2aec9fd9fbd17af6803d75c7f631c1ae3g

### CODE REVIEW

File: src/chunk_dss.c [hunk 1]
```diff
@@ -63,36 +65,40 @@ chunk_dss_prec_set(dss_prec_t dss_prec)
...
 }
 
+static void
+chunk_dss_extending_start(void) {
+	spin_t spinner;
+
+	spin_init(&spinner);
+	while (true) {
+		unsigned expected = 0;
+		if (atomic_cas_u(&dss_extending, expected, 1)) {
```
Comment (interwq) [inline src/chunk_dss.c]: The atomic_cas API here returns true on failure. So we would want to flip the condition. Looks this is causing the Travis test failure.

**Comment by davidtgoldblatt** [comment]
This looks good, but I'm going to delay merging until after the weekend so that we can fix our travis build (they updated some of their packages in a way that introduced a clang/libstdc++ incompatibility it seems).

If, as I suspect, you're interested in this as a fix for #937, I'd regard this commit as safe enough for your purposes, even if we don't cut a release specifically for it.

**Comment by xhochy** [comment]
@davidtgoldblatt this is only a workaround for us for #937. With the 4.x series we don't run into #937 but into #802. With this change, we can for now release Apache Arrow with jemalloc as its main allocator, just keeping it on the 4.x series for now.

In the next weeks; I can have a more in depth look into #937 but I first need to get myself acquainted with the matter.

**Comment by davidtgoldblatt** [comment]
Yes, apologies, that's what I meant. I pinged the thread there with an update.

**Comment by interwq** [comment]
@xhochy: just a nitpick, can you please squash the two commits into one? Thanks.

**Comment by xhochy** [comment]
I squashed the commits all into one. Anything that needs to be done to get this merged?

**Comment by interwq** [comment]
Thanks for the patch @xhochy ! LGTM. Merging.

**Pull request merged**